CI failure: CI — Integration test infrastructure failing after Vitest 4 compatibility changes

## CI Failure Report

**Workflow:** CI
**Branch:** main
**Commit:** 425351cba9f732b16a32a476a678ba727799fd31
**Run:** https://github.com/Action-Llama/action-llama/actions/runs/23393258966

## Failure Summary

Integration tests are systematically failing with agents completing in 'error' state instead of 'completed' state across all test files. This appears to be an infrastructure issue with the integration test harness introduced by recent Vitest 4 compatibility changes.

## Failed Job(s)

### test

**Error output:**
```
AssertionError: expected 'error' to be 'completed' // Object.is equality

Expected: "completed"
Received: "error"

❯ packages/action-llama/test/integration/locks.test.ts:141:30
❯ packages/action-llama/test/integration/rerun.test.ts:40:29
❯ packages/action-llama/test/integration/rlock-logging.test.ts:145:32
❯ packages/action-llama/test/integration/scheduler.test.ts:46:24
❯ packages/action-llama/test/integration/signals.test.ts:26:24
❯ packages/action-llama/test/integration/triggers.test.ts:50:30
❯ packages/action-llama/test/integration/webhooks.test.ts:51:24
```

29 failed integration tests, 1247 passed unit tests.

## Root Cause Analysis

The issue appears to stem from commit 425351c "Fix e2e tests for Vitest 4 compatibility and Docker availability". This commit introduced:

1. **Docker availability checks** - Added `isDockerAvailable()` checks with graceful exit in global setup
2. **Vitest 4 compatibility changes** - Moved `pool` configuration from `test.pool` to top-level
3. **Error handling changes** - Modified how Docker unavailability is handled

The problem is in the logic of the Docker availability checks:

- `packages/e2e/src/global-setup.ts` exits gracefully with `process.exit(0)` when Docker is unavailable
- `packages/e2e/src/harness.ts` throws an error when Docker is unavailable
- Integration tests use `describe.skipIf(!DOCKER)` to skip tests when Docker is unavailable

This creates an inconsistent state where global setup "succeeds" but individual test harnesses fail, or Docker is detected as available but container execution fails.

## Suggested Fix

The Docker availability logic needs to be consistent across the test infrastructure:

1. **Fix global setup logic** - Don't use `process.exit(0)` which can interfere with test execution
2. **Ensure consistent Docker detection** - Use the same Docker availability check across all components
3. **Verify Vitest 4 pool configuration** - Ensure the pool configuration changes are correct
4. **Add better error reporting** - When Docker-related failures occur, provide clearer error messages

**Suggested code changes:**

```diff
// packages/e2e/src/global-setup.ts
export async function setup() {
  const dockerAvailable = await isDockerAvailable();
  if (!dockerAvailable) {
-    console.warn("Docker is not available. E2E tests require Docker to run.");
-    console.warn("Please ensure Docker is installed and running, or run tests in an environment with Docker support.");
-    process.exit(0); // Exit gracefully instead of failing
+    throw new Error("Docker is not available. E2E tests require Docker to run. Please ensure Docker is installed and running.");
  }
```

Also verify that the integration test Docker detection is working correctly and that containers can actually be started in the CI environment.

## Additional Context

Recent commits around this change:
- `425351c` Fix e2e tests for Vitest 4 compatibility and Docker availability (#235)
- `7187099` Reduce secret/file permissiveness in local staging (#230)

This is a different issue from #234 which was about credential permissions. This failure affects the core integration test infrastructure and prevents all integration tests from running successfully.

---
*This issue was automatically created by the gh-actions-responder agent.*

--- a/packages/e2e/src/docker-utils.ts
+++ b/packages/e2e/src/docker-utils.ts
@@ -1,0 +1,26 @@
+import Docker from "dockerode";
+
+/**
+ * Check if Docker daemon is available by attempting to ping it.
+ * This provides consistent Docker availability detection across all e2e components.
+ */
+export async function isDockerAvailable(): Promise<boolean> {
+  try {
+    const docker = new Docker();
+    await docker.ping();
+    return true;
+  } catch (error) {
+    return false;
+  }
+}
+
+/**
+ * Assert that Docker is available, throwing a consistent error message if not.
+ * Use this in components that require Docker to be available.
+ */
+export async function assertDockerAvailable(): Promise<void> {
+  const dockerAvailable = await isDockerAvailable();
+  if (!dockerAvailable) {
+    throw new Error("Docker is not available. E2E tests require Docker to run. Please ensure Docker is installed and running.");
+  }
+}

--- a/packages/e2e/src/global-setup.ts
+++ b/packages/e2e/src/global-setup.ts
@@ -1,23 +1,9 @@
 import Docker from "dockerode";
-
-async function isDockerAvailable(): Promise<boolean> {
-  try {
-    const docker = new Docker();
-    await docker.ping();
-    return true;
-  } catch (error) {
-    return false;
-  }
-}
+import { assertDockerAvailable, isDockerAvailable } from "./docker-utils.js";
 
 export async function setup() {
   // Check if Docker is available before proceeding
-  const dockerAvailable = await isDockerAvailable();
-  if (!dockerAvailable) {
-    console.warn("Docker is not available. E2E tests require Docker to run.");
-    console.warn("Please ensure Docker is installed and running, or run tests in an environment with Docker support.");
-    process.exit(0); // Exit gracefully instead of failing
-  }
+  await assertDockerAvailable();
 
   const docker = new Docker();
   

--- a/packages/e2e/src/harness.ts
+++ b/packages/e2e/src/harness.ts
@@ -4,6 +4,7 @@ import { generateKeyPairSync } from "crypto";
 import { promises as fs } from "fs";
 import path from "path";
 import { Client as SSHClient } from "ssh2";
+import { assertDockerAvailable, isDockerAvailable } from "./docker-utils.js";
 
 export interface ContainerInfo {
   id: string;
@@ -25,21 +26,9 @@ export class E2ETestContext {
     this.tempDir = `/tmp/e2e-${this.runId}`;
   }
 
-  async isDockerAvailable(): Promise<boolean> {
-    try {
-      await this.docker.ping();
-      return true;
-    } catch (error) {
-      return false;
-    }
-  }
-
   async setup() {
     // Check if Docker is available
-    const dockerAvailable = await this.isDockerAvailable();
-    if (!dockerAvailable) {
-      throw new Error("Docker is not available. E2E tests require Docker to run. Please ensure Docker is installed and running.");
-    }
+    await assertDockerAvailable();
 
     // Create temp directory for test artifacts
     await fs.mkdir(this.tempDir, { recursive: true });

